### DISCUSSION TITLE
test warnings in docstrings without monkey-patching warnings.showwarning

### DISCUSSION THREAD

**Discussion by dimpase** [pytest-dev/pytest#10565] [General]
In order to test warnings in docstrings of functions and classes, it appears necessary to monkey-patch `warnings.showwarning`,
e.g. as follows
```python
# c.py
import warnings
dowarn = warnings.showwarning
def showwarn(message, category, filename, lineno, file=None, line=None):
    import sys
    dowarn(message, category, filename, lineno, sys.stdout, line)

warnings.showwarning = showwarn # use stdout instead of stderr


class tst:
    """
    >>> import warnings
    >>> warnings.warn("foo")
    <doctest ...: UserWarning: foo
      warnings.warn("foo")
    """
    pass
```

Then `pytest --doctest-modules -p no:warnings c.py` passes - otherwise it does not get any output of `warn()` call, and the test fails.

Is there a more standard way to accomplish this?  Obviously monkey-patching of the standard library is not what one should normally do.

**Comment by The-Compiler** [comment]
I'm not too familiar with doctests, but what about turning warnings into errors (`pytest -Werror` or `filterwarnings = error` in the config), and then testing for the exception?

```python
class tst:
    """
    >>> import warnings
    >>> warnings.warn("foo")
    Traceback (most recent call last):
        ...
    UserWarning: foo
    """
    pass
```

**Reply by dimpase** [reply]
> I'm not too familiar with doctests, but what about turning warnings into errors (`pytest -Werror` or `filterwarnings = error` in the config), and then testing for the exception?

We have many thousands of doctests, many of these are meant to produce deprecation warnings, experimental status warnings, etc etc. It seems this kind of change would require massive changes in our [sources](https://github.com/sagemath/sage)

**Reply by The-Compiler** [reply]
You could also use [`warnings.catch_warnings()`](https://docs.python.org/3/library/warnings.html#warnings.catch_warnings) and then check the warnings programatically. Also see the relevant Python docs around [testing warnings](https://docs.python.org/3/library/warnings.html#testing-warnings).
